### ISSUE TITLE
Option to re-clone old primary if use_pg_rewind is false

### ISSUE THREAD

**Issue by mbanck** [zalando/patroni#941]
If `use_pg_rewind` is set to `false`, then a former primary will not get back online after having added local changed before failover (e.g. during a network outage).

In that case, the log says (sorry for the german, but you get the idea):

```
2019-01-22 12:44:20.836 UTC [8486] LOG:  neue Zeitleiste 2 zweigte von der aktuellen Datenbanksystemzeitleiste 1 vor dem aktuellen Wiederherstellungspunkt 0/5000098 ab
2019-01-22 12:44:25.838 UTC [8515] FATAL:  konnte WAL-Streaming nicht starten: FEHLER:  angeforderter Startpunkt 0/5000000 auf Zeitleiste 1 ist nicht in der History dieses Servers
	DETAIL:  Die History dieses Servers zweigte von Zeitleiste 1 bei 0/4000060 ab.
```

However, `patronictl` and `journalctl` think not a lot is wrong:

```patronictl -c /etc/patroni/9.6-main.yml list                                                           kohn: Tue Jan 22 13:46:22 2019

+----------+--------+------------+--------+---------+-----------+
| Cluster  | Member |    Host    |  Role  |  State  | Lag in MB |
+----------+--------+------------+--------+---------+-----------+
| 9.6-main |  pg1   | 10.0.3.203 | Leader | running |         0 |
| 9.6-main |  pg2   | 10.0.3.89  |        | running |           |
| 9.6-main |  pg3   | 10.0.3.204 |        | running |         0 |
+----------+--------+------------+--------+---------+-----------+
```

```
Jan 22 12:44:06 pg2 patroni[8284]: 2019-01-22 12:44:06,801 INFO: Lock owner: pg1; I am pg2
Jan 22 12:44:06 pg2 patroni[8284]: 2019-01-22 12:44:06,801 INFO: does not have lock
Jan 22 12:44:06 pg2 patroni[8284]: 2019-01-22 12:44:06,801 INFO: establishing a new patroni connection to the postgres cluster
Jan 22 12:44:06 pg2 patroni[8284]: 2019-01-22 12:44:06,810 INFO: no action.  i am a secondary and i am following a leader
Jan 22 12:44:13 pg2 patroni[8284]: 2019-01-22 12:44:13,491 INFO: Lock owner: pg1; I am pg2
Jan 22 12:44:13 pg2 patroni[8284]: 2019-01-22 12:44:13,491 INFO: does not have lock
Jan 22 12:44:13 pg2 patroni[8284]: 2019-01-22 12:44:13,498 INFO: no action.  i am a secondary and i am following a leader
```

The missing `0` in `Lag in MB` might be a hint, but it seems to me that (i) patroni itself should know better that itself (`pg2`) is not really following the leader and (ii) it would be nice if there was an option to simply re-clone the former primary for cases where there is not a lot of data in the cluster and it should be fast.

**Comment by CyberDem0n** [comment]
Hi @mbanck,

We always consider the data more important than anything else and therefore Patroni never deletes data directory unless it was explicitly told to do so (`patronictl reinit` or when `remove_data_directory_on_rewind_failure` is set to true).

Right now, if pg_rewind is disabled (actually it is not enabled by default) or can't be used for some reason, Patroni is not even trying to figure out that there is some data loss and the former master can't be converted to the replica. In your case timeline 2 was created at `0/4000060`, while the former master has been at `0/5000000`. Unfortunately, postgres doesn't provide good mechanisms to check such kind of failures (`new timeline 2 forked off current database system timeline 1 before current recovery point 0/5B000098`) and we have to do it manually, i.e. get history file from the master and compare the branching point with the current LSN position on the former master. The last one either requires clean shutdown (`Database cluster state: shut down or shut down in recovery`) or postgres being running. Since the former master probably crashed, chances that shut down was clean are very-very small.

So basically the only Patroni can do right now, create recovery.conf and try to start postgres. Depending on the state of PGDATA postgres might:
1. refuse to start (abort start at some moment). That usually happens if the exclusive backup was taken during the moment when it crashed.
2. stuck in the situation were it doesn't accept connections (and not streaming either)
3. start streaming from the master, but refuse connections (strange, but we have seen it in the wild)
4. start accepting connections but dosen't stream from the master.
5. start streaming from the master and become absolutely healthy replica

Cases 1 and 4 we can detect and maybe reinitialize replica if the special setting is enabled, but 2 and 3 Patroni can't really handle, because it requires looking into postgres logs.

**Comment by CyberDem0n** [comment]
> Did you consider storing the primary history in DCS?

It is already stored there. But it happened after we implemented fetching history file from the master via replication protocol: https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L1338

> patroni wouldn't need to query remote postgres instances for info

Well, in this case it uses a replication connection, so it doesn't change pretty much anything.

> I guess Patroni fails to get the history info and does not get to pg_rewind

Nope, before fetching the history file it wants to verify that postgres is already running as master and execute CHECKPOINT (https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L1326). Executing pg_rewind right after promotion (when the new master didn't managed to create a new checkpoint) gives unpredictable results.

> so remove_data_directory_on_rewind_failure isn't taken into consideration either

Yes, because in fact that's not a pg_rewind failure, but Patroni can't access the master with superuser credentials. There are bunch of reasons why pg_rewind might fail: master is not accessing, or it is still in recovery, or it didn't finish checkpoint yet and we really want to eliminate all of them, so failure of pg_rewind will have a single meaning, if it failed, there is nothing else we can do. In your case it is one of the pre-checks fails, therefore it doesn't take remove_data_directory_on_rewind_failure into consideration.


> so I think it would be useful to mention this in the patronictl list output somehow, maybe via "not streaming" in the Lag in MB field?

Streaming/not streaming might be hard, `pg_stat_wal_receiver` view was introduced only in 9.6, so the only way to get this information is fetching it from the master or cascading replica...
What we can easily do - report the current timeline.

In general, I think we could implement a new switch, which will work similarly to the remove_data_directory_on_rewind_failure but when pg_rewind isn't enabled at all, but I am not sure when I'll have time for that.

**Comment by mbanck** [comment]
> > Did you consider storing the primary history in DCS?
> 
> It is already stored there. But it happened after we implemented fetching history file from the master via replication protocol: https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L1338

I see, ok.

> > patroni wouldn't need to query remote postgres instances for info
> 
> Well, in this case it uses a replication connection, so it doesn't change pretty much anything.

D'oh ok.
 
> > I guess Patroni fails to get the history info and does not get to pg_rewind
> 
> Nope, before fetching the history file it wants to verify that postgres is already running as master and execute CHECKPOINT (https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L1326). Executing pg_rewind right after promotion (when the new master didn't managed to create a new checkpoint) gives unpredictable results.

OK, though it seems to do the `check_leader_is_not_in_recovery` call before checking whether rewind is even needed. I wonder whether the `IDENTIFY_SYSTEM` call would still work and the `check_leader_is_not_in_recovery` call could be pushed to the beginning of the `rewind` function - then all the superuser-needing connections are in that function. This would also mean that the information that rewind is necessary could be gathered.

> > so remove_data_directory_on_rewind_failure isn't taken into consideration either
> 
> Yes, because in fact that's not a pg_rewind failure, but Patroni can't access the master with superuser credentials. There are bunch of reasons why pg_rewind might fail: master is not accessing, or it is still in recovery, or it didn't finish checkpoint yet and we really want to eliminate all of them, so failure of pg_rewind will have a single meaning, if it failed, there is nothing else we can do. In your case it is one of the pre-checks fails, therefore it doesn't take remove_data_directory_on_rewind_failure into consideration.

I see. 

As an aside, shouldn't failure to run `self.checkpoint(r)` in https://github.com/zalando/patroni/blob/master/patroni/postgresql.py#L1395 set `self._rewind_state` to `REWIND_STATUS.FAILED` as well before returning?

> > so I think it would be useful to mention this in the patronictl list output somehow, maybe via "not streaming" in the Lag in MB field?
> 
> Streaming/not streaming might be hard, `pg_stat_wal_receiver` view was introduced only in 9.6, so the only way to get this information is fetching it from the master or cascading replica...
> What we can easily do - report the current timeline.

Yeah, maybe "not streaming" is not so great, but something like "old timeline", "rewind required" or maybe even "broken" would be a good hint that manual intervention is needed; I think this is just not clear enough right now just from looking at `patronictl list` output. I have opened a new issue (#942) for that.
 
> In general, I think we could implement a new switch, which will work similarly to the remove_data_directory_on_rewind_failure but when pg_rewind isn't enabled at all, but I am not sure when I'll have time for that.

Ok, thanks.
